crates/network-monitor/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::time::Duration;
+use std::{fs, io, mem};
+
+use camino::Utf8PathBuf;
+use network_scanner_net::runtime::Socket2Runtime;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use time::UtcDateTime;
+use tokio::select;
+use tokio_util::sync::CancellationToken;
+use tracing::warn;
+
+use network_scanner::ping;
+
+mod log_queue;
+mod state;
+
+pub use crate::state::State;
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub enum SetConfigError {
+    Io(#[from] io::Error),
+    Serde(#[from] serde_json::Error),
+}
+
+pub async fn set_config(config: MonitorsConfig, state: Arc<State>) -> Result<(), SetConfigError> {
+    let file = fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(&state.cache_path)?;
+
+    let mut config_write = state.config.write().await;
+
+    serde_json::to_writer_pretty(&file, &config)?;
+
+    let old_config = mem::replace(&mut *config_write, config);
+
+    let new_config_set: HashSet<MonitorDefinition> = config_write.monitors.clone().into_iter().collect();
+    let old_config_set: HashSet<MonitorDefinition> = old_config.monitors.clone().into_iter().collect();
```

Comment:
issue: Drop the lock as soon as you are done with it, so that it is released earlier.